pymatsolver/mumps/_init_.py

```diff
@@ -4,162 +4,228 @@
+        -800: "Temporary error associated to the current MUMPS release, subject to change or disappearance in the future. If INFO(2)=5, then this error is due to the fact that the elemental matrix format (ICNTL(5)=1) is currently incompatible with a BLR factorization (ICNTL(35) != 0).",
+    }
+
+    warning_messages = {
+        1: "Index (in IRN or JCN) out of range. Action taken by subroutine is to ignore any such entries and continue. INFO(2) is set to the number of faulty entries. Details of the first ten are printed on unit ICNTL(2).",
+        2: "During error analysis the max-norm of the computed solution was found to be zero.",
+        4: "User data JCN has been modified (internally) by the solver.",
+        8: "Warning return from the iterative refinement routine. More than ICNTL(10) iterations are required.",
+    }
+
+    if ierr == 0:
+        return "No error occurred."
+    elif ierr < 0:
+        return error_messages.get(ierr, "An unknown error occurred.")
+    else:
+        # The warning values are actually bitwise flags.
+        messages = []
+        for bit in warning_messages.keys():
+            if ierr & bit != 0:
+                messages.append(warning_messages[bit])
+
+        if not messages:
+            messages.append("An unknown warning occurred.")
+
+        return "\n".join(messages)
+
+
+class _Pointer:
+    """Store a pointer with a destroy function that gets called on garbage collection.
+
+    Multiple Mumps solver instances can point to the same factor in memory.
+    This class ensures that the factor is garbage collected only when all references are removed.
+
+    You can always clean the Solver, and it will explicitly call the destroy method.
     """
-    def __init__(self, pointerINT, destroyCall):
+    def __init__(self, pointerINT, dtype):
         self.INT = pointerINT
-        self.destroyCall = destroyCall
+        # The destroy function should only be called from this class upon deallocation.
+        if dtype == float:
+            self._destroy_func = _MUMPSINT.destroy_mumps
+        elif dtype == complex:
+            self._destroy_func = _MUMPSINT.destroy_mumps_cmplx
+        else:
+            raise ValueError(f"Attempted to use an invalid data type ({dtype})")
 
     def __del__(self):
-        self.destroyCall(self.INT)
+        self._destroy_func(self.INT)
+
 
 class Mumps(Base):
     """
 
     documentation::
 
-        http://mumps.enseeiht.fr/doc/userguide_4.10.0.pdf
+        https://mumps-solver.org/doc/userguide_5.6.1.pdf
 
     """
 
     transpose = False
-    symmetric = False
 
     @property
     def T(self):
+        properties_with_setters = []
+        for a in dir(self.__class__):
+            attr = getattr(self.__class__, a)
+            if hasattr(attr, "fset") and attr.fset is not None:
+                properties_with_setters.append(a)
+        kwargs = {attr: getattr(self, attr) for attr in properties_with_setters}
+
         newMS = self.__class__(
```

Comment:
I think the point before was that mumps could use a factorization of `A` to solve systems with `A.T` instead of refactoring the entire matrix, as like what will happen here (This might not be the case though.)

Reply:
Nice catch - and yes, that's exactly the point here.  Fixed in c848dd8427f743f8d2fd9290ddd1c0ec4f008ac8.

I must have introduced this while debugging why the transpose was resulting in a Hermitian transpose for complex matrices (which is not how the pardiso or the rest of the solvers handle this case).

The following are now equivalent but `AinvT1` doesn't need a refactorization:
```python
AinvT1 = Mumps(A).T
AinvT2 = Mumps(A.T)
```

This behavior matches all the other solvers.

I contemplated adding a `.H` (Hermitian transpose) operator such that:

```python
Mumps(A).H == Mumps(A.T.conjugate())
```

However, this presents a floating-point +0 / -0 problem that I'm not thrilled about.  Given the matrix
```
[ 1 + 0j,  0 + 0j ]
[ 0 + 0j,  1 + 1j ]
```
the inverse is

```
[ 1 + 0j, 0   +   0j ]
[ 0 + 0j, 0.5 - 0.5j ]
```

Since Mumps doesn't take the Hermitian transpose itself, we'd need to emulate that behavior by conjugating the RHS, telling Mumps to solve the transpose problem, and then conjugating again.

The inverse of the matrix's conjugate transpose (i.e. `Mumps(A.T.conjugate())` is
```
[ 1 + 0j, 0   +   0j ]
[ 0 + 0j, 0.5 + 0.5j ]
```
as computed by Mumps, scipy, or pardiso.

However, the inverse as computed with the emulation (i.e. `Mumps(A).H`) is
```
[ 1 - 0j, 0   -   0j ]
[ 0 - 0j, 0.5 + 0.5j ]
```

That's a little iffy for me.  I'd rather have the solvers all behave the same way on simple cases.

Reply:
Honestly, I'm not concerned about the differences between +0 and -0 (A floating point thing), I don't think we're ever doing anything that divides by 0 which would be the only place that makes a difference.

Reply:
That has me convinced.

I went away from the `.H` property because the scipy.sparse API doesn't have that property and I wanted to present the same interface here.  To take the Hermitian transpose of a scipy.sparse matrix `A`, you run `A.T.conjugate()`, so I added a conjugate operator here.  I also added unit tests to ensure that `Mumps(A)T.conjugate()` gives the same results as `Mumps(A.T.conjugate())`.